File: 48.rotate-image.js
/*
 * @lc app=leetcode id=48 lang=javascript
 *
 * [48] Rotate Image
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function (matrix) {
  const n = matrix.length;
  for (let i = 0; i < n - 1; ++i) {
    for (let j = i + 1; j < n; ++j) {
      const tmp = matrix[i][j];
      matrix[i][j] = matrix[j][i];
      matrix[j][i] = tmp;
    }
  }
  for (let i = 0; i < n; ++i) {
    for (let j = 0; j < n / 2; ++j) {
      const tmp = matrix[i][j];
      matrix[i][j] = matrix[i][n - 1 - j];
      matrix[i][n - 1 - j] = tmp;
    }
  }
};
// @lc code=end

File: 48.rotate-image.py
#
# @lc app=leetcode id=48 lang=python3
#
# [48] Rotate Image
#

# Time:  O(n^2)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)

        def swapMatrix(i1, j1, i2, j2):
            matrix[i1][j1], matrix[i2][j2] = matrix[i2][j2], matrix[i1][j1]

        for i in range(n - 1):
            for j in range(n - i):
                swapMatrix(i, j, n - 1 - j, n - 1 - i)
        for i in range(n // 2):
            for j in range(n):
                swapMatrix(i, j, n - 1 - i, j)
        return

        for i in range(n // 2):
            for j in range(i, n - 1 - i):
                tmp = matrix[i][j]
                matrix[i][j] = matrix[n - 1 - j][i]
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j]
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i]
                matrix[j][n - 1 - i] = tmp
        return


# @lc code=end
